Readme.md
<div align="center">
  <img src="./docs/logo.png" alt="StormDB logo">

  <p>🌩️ StormDB is a tiny, lightweight, 0 dependency, easy-to-use JSON-based database that allows users to quickly and easily achieve data persistence by provided an engine to store and access JSON data, for NodeJS the browser or Electron.</p>

  <a href="https://npmjs.com/package/stormdb">
    <img src="https://img.shields.io/npm/v/stormdb?color=brightgreen">
  </a>
  <a href="https://bundlephobia.com/result?p=stormdb">
    <img src="https://img.shields.io/bundlephobia/minzip/stormdb?color=brightgreen">
  </a>
  <img src="https://img.shields.io/badge/dependencies-0-brightgreen?color=blue">
  <a href="./LICENSE">
    <img src="https://img.shields.io/badge/license-MIT-blue">
  </a>
</div>

<br>

> Example: Add a post entry under users.tom and save it to the database.

```js
db.get("users")
  .get("tom")
  .push({ title: "Post 1" })
  .save();
```

## Features

- 🏎️ [Blazingly Fast Speeds](./docs/speed.md) - Fast read and write speeds, even when handling large data.
- 📦 [Tiny Size](https://bundlephobia.com/result?p=stormdb) - Tiny source code size allows for blazingly fast loading when speed matters.
- ⚡️ Versatile - Can be used with NodeJS, in the browser or in Electron.

## Usage

Install StormDB through NPM:

```
$ npm i stormdb
```

Basic usage with NodeJS:

```js
const StormDB = require("stormdb");

// start db with "./db.stormdb" storage location
const engine = new StormDB.localFileEngine("./db.stormdb");
const db = StormDB(engine);

// set default db value if db is empty
db.default({ users: [] });

// add new users entry
db.get("users").push({ name: "tom" });

// update username of first user
db.get("users")
  .get(0)
  .get("name")
  .set("jeff");

// save changes to db
db.save();
```

The `db.stormdb` database file is updated to:

```js
{
  "users": [
    {"name":"jeff"}
  ]
}
```

StormDB is designed to be flexible, and can be used in NodeJS, the browser or even Electron with very small adaptations to the code. Examples usages can be seen below:

- [Browser Usage](./examples/browser.md)
- [NodeJS Server](./examples/node.md)

## Engine API

For expanding functionality, each database initialized can be expanded with the following options, in the format `new Engine(path, options);`.

- `serialize` - function to serialize data before writing it to the database.
- `deserialize` - function to deserialize data from the database.

## Database Operations Examples

Change Value of Key in Database:

```js
db.get("old").set("newData");
// before: {"old": "oldData"}
// after: {"old": "newData"}
```

Set Key-Value Pair on Dictionary Property:

```js
db.set("key", "value").save();
// before: {}
// after: {"key": "value"}
```

Delete Value:

```js
db.delete("key");
// before: {'key': 'value', 'key2': 'value2'}
// after: {'key2': 'value2'}
```

Set Key-Value Pair on Dictionary with Shorthand Syntax:

```js
db.set("key.key2", "value").save();
// before: {}
// after: {"key": {"key2": "value"}}
```

Set Default Data for Empty Database:

```js
db.default({ name: "tom" });

// actual db: {}
console.log(db.get("name")); // prints "tom"
```

Push Item to Array Property:

```js
db.get("list")
  .push(1)
  .save();

// before: {'list': []}
// after: {'list': [1]}
```

Filter All Elements under 5:

```js
// before = {'list': [1,2,6,1]}
// output = {'list': [6]}

// get list from db
let value = db.get("list").value();

// delete all elements with value under 5
value.forEach(function(el, i) {
  if (el < 5) {
    db.get("list")
      .get(i)
      .delete();
  }
});

// remove all null elements
value = value.filter(el => el !== null);
db.get("list").set(value);

// save db
db.save();
```

Change Element with Highest Value:

```js
// before = {'users': [{value: 10}, {value: 5}, {value: 6}]}
// after = {'users': [{value: "changed"}, {value: 5}, {value: 6}]}

// get list from db
let values = db.get("users").value();

// sort list with highest value first
values = values.sort((a, b) => b.value - a.value);

// change value of highest element
values[0]["value"] = "changed";

// replace list with new list with changed values
db.get("users").set(values);

// save db
db.save();
```

Map List, Squaring Each Number in List:

```js
// before = {'data': [1,2,3,4,5]}
// after = {'data': [1,4,9,16,25]}

// get values list
let values = db.get("data").value();

// map list, squaring each number
values = values.map(x => x ** 2);

// reassign list to db
db.get("data").set(values);

// save db
db.save();
```

Leverage Serialize and Deserialize functions to encrypt and decrypt data:

```js
const engine = new StormDB.localFileEngine("./db.stormdb", {
  serialize: data => {
    // ecrypt and serialize data
    return encrypt(JSON.stringify(data));
  },
  deserialize: data => {
    // decrypt and deserialize data
    return JSON.parse(decrypt(data));
  }
});
const db = StormDB(engine);
```

## Credit

Author: [Tom](https://github.com/TomPrograms)

## License

[MIT](LICENSE)


Changelog.md
# 0.1.1 - 20/02/2020

### Added

- NPM keywords.

### Changed

- Improved Readme - added more documentation and improved badges.

### Fixed

- Inaccurate changelog information.

# 0.1.0 - 20/02/2020

- 🎉 Initial Release of StormDB.


src/stormdb.js
class StormDB {
  constructor(engine, options = {}) {
    this.engine = engine;

    this.state = this.engine.init();
    this.pointers = [];
  }

  default(defaultValue) {
    let stateEmpty =
      Object.keys(this.state).length === 0 && this.state.constructor === Object;

    if (stateEmpty) this.state = defaultValue;

    return this;
  }

  length() {
    this.pointers.push("length");
    return this;
  }

  delete() {
    let enclosing = this.state;
    for (let i = 0; i < this.pointers.length - 1; i++) {
      enclosing = enclosing[this.pointers[i]];
    }

    let final = this.pointers[this.pointers.length - 1];
    delete enclosing[final];
  }

  push(value) {
    let list = this.value();

    if (!Array.isArray(list)) throw new Error("You can only push to lists.");

    list.push(value);
    this.set(list);

    return this;
  }

  get(value) {
    let clone = Object.assign(Object.create(Object.getPrototypeOf(this)), this);

    clone.pointers = [...clone.pointers];
    clone.pointers.push(value);

    return clone;
  }

  set(key, value) {
    if (value === undefined) {
      this.setValue(key);
    } else {
      let extraPointers;
      if (typeof key === "string") extraPointers = key.split(".");
      else extraPointers = [key];

      this.setValue(value, extraPointers);
    }
    return this;
  }

  value() {
    let data = this.state;
    for (let i = 0; i < this.pointers.length; i++) {
      data = data[this.pointers[i]];
    }

    return data;
  }

  setValue(value, pointers = [], setrecursively = true) {
    let depth = 0;

    pointers = [...this.pointers, ...pointers];

    const func = (a, b) => {
      depth += 1;

      let finalLevel = depth === pointers.length;
      if (setrecursively && typeof a[b] === "undefined" && !finalLevel) {
        a[b] = {};
        return a[b];
      }

      if (finalLevel) {
        a[b] = value;
        return value;
      } else {
        return a[b];
      }
    };
    pointers.reduce(func, this.state);
  }

  save() {
    return this.engine.write(this.state);
  }
}

module.exports = function(engine, options = {}) {
  return new StormDB(engine, options);
};


src/engine/local.js
const Base = require("./base.js");
const fs = require("fs");

module.exports = class LocalEngine extends Base {
  constructor(path, options = {}) {
    super();

    this.path = path;

    this.async = options.async || false;
    this.serialize = options.serialize || JSON.stringify;
    this.deserialize = options.deserialize || JSON.parse;
  }

  init() {
    const exists = fs.existsSync(this.path);
    if (!exists) {
      fs.writeFileSync(this.path, this.serialize({}));
      return {};
    } else {
      return this.read();
    }
  }

  read() {
    let data = fs.readFileSync(this.path, "UTF-8");
    if (data === "") data = "{}";

    try {
      let json = this.deserialize(data);
      return json;
    } catch (error) {
      error.message =
        "Failed to load StormDB database file - invalid or corrupted format.";
      throw error;
    }
  }

  write(data) {
    // if async, return promise wrapper around async writefile
    if (this.async) {
      return new Promise(
        function(resolve, reject) {
          fs.writeFile(this.path, this.serialize(data), function() {
            resolve();
          });
        }.bind(this)
      );
    }

    fs.writeFileSync(this.path, this.serialize(data));
    return null;
  }
};


src/engine/browser.js
const Base = require("./base.js");

module.exports = class LocalEngine extends Base {
  constructor(path, options = {}) {
    super();

    this.path = path;

    this.serialize = options.serialize || JSON.stringify;
    this.deserialize = options.deserialize || JSON.parse;
  }

  init() {
    const exists = localStorage.getItem(this.path);
    if (!exists) {
      localStorage.setItem(this.path, this.serialize({}));
      return {};
    } else {
      return this.read();
    }
  }

  read() {
    let data = localStorage.getItem(this.path);

    try {
      let json = this.deserialize(data);
      return json;
    } catch (error) {
      error.message =
        "Failed to load StormDB database file - invalid or corrupted format.";
      throw error;
    }
  }

  write(data) {
    localStorage.setItem(this.path, this.serialize(data));
  }
};


benchmarks/writeBenchmark.js
const StormDB = require("../index.js");
const fs = require("fs");

const path = "./db.stormdb";

function parseToDP(num, dp) {
  return num.toFixed(dp);
}

function deleteDB(path) {
  fs.unlinkSync(path);
}

function getFilesizeInBytes(filename) {
  var stats = fs.statSync(filename);
  var fileSizeInBytes = stats["size"];
  return fileSizeInBytes / 1024 / 1024;
}

function runBenchmark(numberToInsert) {
  if (fs.existsSync(path)) deleteDB(path);

  const engine = new StormDB.localFileEngine(path);
  const db = StormDB(engine);

  /* start of core speed test */
  let start = +new Date();

  for (let i = 0; i < numberToInsert; i++) {
    db.set(i, "val");
  }
  db.save();

  let end = +new Date();
  /* end of core speed test */

  let time = end - start;
  let timeInSecs = time / 1000;
  let sizeOfDB = getFilesizeInBytes(path);

  let speedPerSec = parseToDP(numberToInsert / timeInSecs, 2);

  console.log(`Number Wrote: ${numberToInsert}.
  Time Elapsed: ${time}ms
  Speed: ${speedPerSec}/sec
  Size of Data Write: ${parseToDP(sizeOfDB, 2)}mb\n`);
}

console.log("Write Benchmarks:\n");
let numbers = [100000, 1000000, 10000000];
numbers.forEach(val => {
  runBenchmark(val);
});


benchmarks/readBenchmark.js
const StormDB = require("../index.js");
const fs = require("fs");

const path = "./db.stormdb";

function parseToDP(num, dp) {
  return num.toFixed(dp);
}

function deleteDB(path) {
  fs.unlinkSync(path);
}

function createDatabase(num) {
  const engine = new StormDB.localFileEngine(path);
  const db = StormDB(engine);

  for (let i = 0; i < num; i++) {
    db.set(i, "val");
  }
  db.save();
}

function getFilesizeInBytes(filename) {
  var stats = fs.statSync(filename);
  var fileSizeInBytes = stats["size"];
  return fileSizeInBytes / 1024 / 1024;
}

function runBenchmark(numberToInsert) {
  if (fs.existsSync(path)) deleteDB(path);
  createDatabase(numberToInsert);

  /* start of core speed test */
  let start = +new Date();

  const engine = new StormDB.localFileEngine(path);
  const db = StormDB(engine);

  let end = +new Date();
  /* end of core speed test */

  let time = end - start;
  let timeInSecs = time / 1000;
  let sizeOfDB = getFilesizeInBytes(path);

  let speedPerSec = parseToDP(numberToInsert / timeInSecs, 2);

  console.log(`Number Read: ${numberToInsert}.
  Time Elapsed: ${time}ms
  Speed: ${speedPerSec}/sec
  Size of Data Write: ${parseToDP(sizeOfDB, 2)}mb\n`);
}

console.log("Read Benchmarks:\n");
let numbers = [100000, 1000000, 10000000];
numbers.forEach(val => {
  runBenchmark(val);
});
